added custom JSON Encoder

vabene1111 opened the pull request:
This is a first draft to address json exporting as discussed in #619

It appears that if you want to write a custom method to serialize a python object you need to use subclass `JSONEncoder`.

The code proposed solution allows the user to easily dump a json object of a scraper with all methods that do not error out by running
```python
json.dumps(scraper, cls=ScraperJSONEncoder, indent=True)
```

one thing that is not perfect is that the custom encoder is part of a protected module `_utils.py` but I wasn't sure where else to put it. 

One alternative would be to put the method into the `AbstractScraper` as a "toJSON" function, not sure which one is more intuitive/clean.

cydanil commented:
I'm interested in this feature, and I followed the discussion in the linked issue, and I haven't tried these changes yet (but I'm excited to 😉 )

Would it make sense to export to a [schema.org](https://schema.org/Recipe) format? This would probably suit the needs of people exporting recipes back to the web.

jayaddison commented:
> It appears that if you want to write a custom method to serialize a python object you need to use subclass `JSONEncoder`.

This is _mostly_ true, although there is a workaround possible: we could make the [`AbstractScraper`](https://github.com/hhursev/recipe-scrapers/blob/ccda81def2ecb8aabefc7e34f164842ebb060e3a/recipe_scrapers/_abstract.py#L20) inherit from `dict`.

Apparently doing that can [cause some problems](https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/).  I think it's _much_ nicer than requiring custom JSON encoders, personally (it always seems strange to me that JSON serialization of objects isn't easier in Python).

> Would it make sense to export to a [schema.org](https://schema.org/Recipe) format? This would probably suit the needs of people exporting recipes back to the web.

It sorta does - that way, the code here could be used within a browser plugin to cleanup/fix recipe content while navigating the web.  There's some risk though - it might, for example, lower the level-of-effort barrier for spammy sites to re-publish existing recipes without proper attribution.

vabene1111 commented:
Good question. Yesterday before I implemented the feature I thought about using schema.org but today I totally forgot.

Maybe two different versions 🤔

jayaddison commented on recipe_scrapers/_utils.py:
```diff
@@ -175,3 +176,25 @@ def change_keys(obj, convert):
+            for method in dir(o)
+            if callable(getattr(o, method))
+            if not method.startswith("_") and method not in ["soup", "links"]
+        ]
```
Could you add a test case that checks the list of public methods reported for a scraper object?

vabene1111 replied:
i think we should add some test for all of the json serialization and methods used within but I would wait until we have decided how to implement this feature.

jayaddison replied:
Ok.  I disagree and think that writing a test to confirm the list of expected method names would be useful before the serialization implementation.

vabene1111 commented:
I thought a bit more about that and think that we should move the code into a `toJSON()` method of the abstract scraper. This would also allow people to easily expand it with something like a `toMarkdown()` and we can pass parameters to the  function which would allow us to return either a schema.org representation or a method/attribute based one, what do you think ?

jayaddison commented:
> move the code into a toJSON() method of the abstract scraper.

A method sounds good :+1: 

I'd suggest `to_json()` based on this repository's existing [naming style](https://peps.python.org/pep-0008/#descriptive-naming-styles).

vabene1111 commented:
ok so i moved the feature to a function in the abstract scraper.
I also added a general test as requested for the list of public method names even if i think that this kinda defeats the purpose of having a method that dynamically returns everything there is so that you dont have to change anything when updating the scraper, but anyway. 

I sadly do not have the time to implement the additional feature of restructuring the parsed data into a schema.org compliant json. I think its probably best to merge this so the basic function exists and at a later point either me or someone else can simply implement a kwarg to the json function to control if schema.org or method based lists should be returned and add the appropriate code. 

I refrained from adding a test for the json content as that would need to be specific to a parser and in my opinion a developer to the application would not expect a general purpose function to break when changing a parser so to avoid confusion the function does not have a content based test (and should not need to as the content is tested in the parser specific tests)
